### PR TITLE
Editorial : ref extended/modified steps/entries

### LINKED ISSUE #2191: consistently note in algorithms wherever steps may be inserted/modified elsewhere
We have step references now to ensure that algorithm modifications (mostly or wholly from Annex B) refer to the correct step number. But we're inconsistent about actually stating within the algorithm steps that this algorithm is modified elsewhere. At least one example is [B.3.7.2](https://tc39.es/ecma262/#sec-IsHTMLDDA-internal-slot-aec) changing [7.2.15, Abstract Equality Comparison](https://tc39.es/ecma262/#sec-abstract-equality-comparison) with no indication within 7.2.15. Elsewhere we have notes inline as another step. We should do that everywhere.

### PR DESCRIPTION
**Pull request by daemon1024**
Closes #2191

### CODE REVIEW

File: spec.html [hunk 1]
```diff
@@ -5188,6 +5194,9 @@ <h1>Abstract Equality Comparison</h1>
           1. If the mathematical value of _x_ is equal to the mathematical value of _y_, return *true*; otherwise return *false*.
         1. Return *false*.
       </emu-alg>
+      <emu-note>
+	      <p>Additional Equality Comparison steps related to [[IsHTMLDDA]] Internal Slot can be found in <emu-xref href="#sec-IsHTMLDDA-internal-slot-aec"></emu-xref>.</p>
```
Comment (michaelficarra) [inline spec.html:5198]: I would probably prefer for this note to be right in the algorithm steps, just before step 3 where the additional steps are to be inserted. It doesn't have to be an `<emu-note>` when it's inline in the algorithm steps, see [ApplyStringOrNumericBinaryOperator](https://tc39.es/ecma262/#sec-applystringornumericbinaryoperator) step 2 as an example.

@tc39/ecma262-editors Do you agree?

File: spec.html [hunk 2]
```diff
@@ -4197,6 +4200,9 @@ <h1>ToBoolean ( _argument_ )</h1>
             </td>
             <td>
               Return *true*.
+              <emu-note>
+                <p>Alternate algorithm related to [[IsHTMLDDA]] Internal Slot can be found in <emu-xref href="#sec-IsHTMLDDA-internal-slot-to-boolean"></emu-xref>.</p>
```
Comment (michaelficarra) [inline spec.html:4204]: I would phrase it like this:

```suggestion
                <p>An alternate algorithm related to the [[IsHTMLDDA]] internal slot is mandated in section <emu-xref href="#sec-IsHTMLDDA-internal-slot-to-boolean"></emu-xref>.</p>
```

File: spec.html (hunk 1 above)
Comment (ljharb) [reply spec.html:5198]: SGTM

File: spec.html [hunk 3]
```diff
@@ -10697,6 +10706,7 @@ <h2>Syntax</h2>
       SingleLineCommentChar ::
         SourceCharacter but not LineTerminator
     </emu-grammar>
+    <p>A conforming implementation, when parsing source code using the goal symbol |Module|, must not extend, as described in <emu-xref href="#sec-html-like-comments"></emu-xref></p>
```
Comment (ljharb) [inline spec.html:10709]: must not extend what?

File: spec.html [hunk 4]
```diff
@@ -30822,6 +30839,9 @@ <h1>Static Semantics: Early Errors</h1>
             It is a Syntax Error if the List of Unicode code points that is SourceText of |LoneUnicodePropertyNameOrValue| is not identical to a List of Unicode code points that is a Unicode general category or general category alias listed in the &ldquo;Property value and aliases&rdquo; column of <emu-xref href="#table-unicode-general-category-values"></emu-xref>, nor a binary property or binary property alias listed in the &ldquo;Property name and aliases&rdquo; column of <emu-xref href="#table-binary-unicode-properties"></emu-xref>.
           </li>
         </ul>
+        <emu-note>
+          <p>The semantics is extended in <emu-xref href="#sec-patterns-static-semantics-early-errors-annexb"></emu-xref>.</p>
```
Comment (ljharb) [inline spec.html:30843]:
```suggestion
          <p>The semantics are extended in <emu-xref href="#sec-patterns-static-semantics-early-errors-annexb"></emu-xref>.</p>
```

File: spec.html [hunk 5]
```diff
@@ -30836,6 +30856,9 @@ <h1>Static Semantics: CapturingGroupNumber</h1>
           1. Return the Number value for (the MV of |NonZeroDigit| &times;<sub>ℝ</sub> 10<sub>ℝ</sub><sup>_n_</sup> plus the MV of |DecimalDigits|).
         </emu-alg>
         <p>The definitions of &ldquo;the MV of |NonZeroDigit|&rdquo; and &ldquo;the MV of |DecimalDigits|&rdquo; are in <emu-xref href="#sec-literals-numeric-literals"></emu-xref>.</p>
+        <emu-note>
+          <p>The semantics is extended in <emu-xref href="#sec-patterns-static-semantics-early-errors-annexb"></emu-xref>.</p>
```
Comment (ljharb) [inline spec.html:30860]:
```suggestion
          <p>The semantics are extended in <emu-xref href="#sec-patterns-static-semantics-early-errors-annexb"></emu-xref>.</p>
```

File: spec.html [hunk 6]
```diff
@@ -30858,6 +30881,9 @@ <h1>Static Semantics: IsCharacterClass</h1>
         <emu-alg>
           1. Return *true*.
         </emu-alg>
+        <emu-note>
+          <p>The semantics is extended in <emu-xref href="#sec-patterns-static-semantics-is-character-class-annexb"></emu-xref>.</p>
```
Comment (ljharb) [inline spec.html:30885]:
```suggestion
          <p>The semantics are extended in <emu-xref href="#sec-patterns-static-semantics-is-character-class-annexb"></emu-xref>.</p>
```

File: spec.html [hunk 7]
```diff
@@ -31046,6 +31072,9 @@ <h1>Static Semantics: CharacterValue</h1>
           1. Let _ch_ be the code point matched by |IdentityEscape|.
           1. Return the code point value of _ch_.
         </emu-alg>
+        <emu-note>
+          <p>The semantics is extended in <emu-xref href="#sec-patterns-static-semantics-character-value-annexb"></emu-xref>.</p>
```
Comment (ljharb) [inline spec.html:31076]:
```suggestion
          <p>The semantics are extended in <emu-xref href="#sec-patterns-static-semantics-character-value-annexb"></emu-xref>.</p>
```

File: spec.html [hunk 8]
```diff
@@ -31086,6 +31115,9 @@ <h1>Pattern Semantics</h1>
         <p>Patterns are passed to the RegExp constructor as ECMAScript String values in which non-BMP characters are UTF-16 encoded. For example, the single character MUSICAL SYMBOL G CLEF pattern, expressed as a String value, is a String of length 2 whose elements were the code units 0xD834 and 0xDD1E. So no further translation of the string would be necessary to process it as a BMP pattern consisting of two pattern characters. However, to process it as a Unicode pattern UTF16SurrogatePairToCodePoint must be used in producing a List consisting of a single pattern character, the code point U+1D11E.</p>
         <p>An implementation may not actually perform such translations to or from UTF-16, but the semantics of this specification requires that the result of pattern matching be as if such translations were performed.</p>
       </emu-note>
+      <emu-note>
+        <p>The semantics is extended in <emu-xref href="#sec-regular-expression-patterns-semantics"></emu-xref>.</p>
```
Comment (ljharb) [inline spec.html:31119]:
```suggestion
        <p>The semantics are extended in <emu-xref href="#sec-regular-expression-patterns-semantics"></emu-xref>.</p>
```

File: spec.html (hunk 3 above)
Comment (daemon1024) [reply spec.html:10709]: Extend the syntax and semantics. Or should I not mention that and simply add `The semantics and syntax are extended in emu-xref href="#sec-html-like-comments">`

File: spec.html (hunk 1 above)
Comment (daemon1024) [reply spec.html:5198]: I think a more apt reference would be similar to step 13 of [15.1.11](https://tc39.es/ecma262/#sec-globaldeclarationinstantiation) which is to add an additional step telling to replace it with annex b reference but that would need a slight change in [B.3.7.2](https://tc39.es/ecma262/#sec-IsHTMLDDA-internal-slot-aec) i.e. ` The following steps replace step 4 of the Abstract Equality Comparison algorithm: `

Comment (michaelficarra) [reply spec.html:5198]: @daemon1024 I think that would be a great improvement!

File: spec.html [hunk 9]
```diff
@@ -30736,6 +30749,10 @@ <h2>Syntax</h2>
           CharacterEscape[?U]
       </emu-grammar>
 
+      <emu-note>
+	      <p>The syntax is modified and extended in <emu-xref href="#sec-regular-expressions-patterns"></emu-xref>.</p>
```
Comment (michaelficarra) [inline spec.html:30753]: I'd use wording more like this:

```suggestion
	      <p>A number of productions in this section are given alternative definitions in section <emu-xref href="#sec-regular-expressions-patterns"></emu-xref>.</p>
```

File: spec.html [hunk 10]
```diff
@@ -16265,11 +16276,11 @@ <h1>Runtime Semantics: BlockDeclarationInstantiation ( _code_, _env_ )</h1>
             1. If IsConstantDeclaration of _d_ is *true*, then
               1. Perform ! _env_.CreateImmutableBinding(_dn_, *true*).
             1. Else,
-              1. [id="step-blockdeclarationinstantiation-createmutablebinding"] Perform ! _env_.CreateMutableBinding(_dn_, *false*).
+              1. [id="step-blockdeclarationinstantiation-createmutablebinding"] Perform ! _env_.CreateMutableBinding(_dn_, *false*). NOTE : This step is replaced in section <emu-xref href="#sec-web-compat-blockdeclarationinstantiation"></emu-xref>.
           1. If _d_ is a |FunctionDeclaration|, a |GeneratorDeclaration|, an |AsyncFunctionDeclaration|, or an |AsyncGeneratorDeclaration|, then
             1. Let _fn_ be the sole element of the BoundNames of _d_.
             1. Let _fo_ be InstantiateFunctionObject of _d_ with argument _env_.
-            1. [id="step-blockdeclarationinstantiation-initializebinding"] Perform _env_.InitializeBinding(_fn_, _fo_).
+            1. [id="step-blockdeclarationinstantiation-initializebinding"] Perform _env_.InitializeBinding(_fn_, _fo_). NOTE : This step is replaced in section <emu-xref href="#sec-web-compat-blockdeclarationinstantiation"></emu-xref>.
```
Comment (ljharb) [inline spec.html:16283]:
```suggestion
            1. [id="step-blockdeclarationinstantiation-initializebinding"] Perform _env_.InitializeBinding(_fn_, _fo_). NOTE: This step is replaced in section <emu-xref href="#sec-web-compat-blockdeclarationinstantiation"></emu-xref>.
```
(please remove the space between "NOTE" and ":" throughout)

File: spec.html [hunk 11]
```diff
@@ -10697,6 +10704,7 @@ <h2>Syntax</h2>
       SingleLineCommentChar ::
         SourceCharacter but not LineTerminator
     </emu-grammar>
+    <p>The semantics and syntax are extended in <emu-xref href="#sec-html-like-comments"></emu-xref></p>
```
Comment (michaelficarra) [inline spec.html:10707]: I'd prefer the same wording I suggested above here as well:

```suggestion
    <p>A number of productions in this section are given alternative definitions in section <emu-xref href="#sec-html-like-comments"></emu-xref></p>
```

File: spec.html [hunk 12]
```diff
@@ -30822,6 +30837,9 @@ <h1>Static Semantics: Early Errors</h1>
             It is a Syntax Error if the List of Unicode code points that is SourceText of |LoneUnicodePropertyNameOrValue| is not identical to a List of Unicode code points that is a Unicode general category or general category alias listed in the &ldquo;Property value and aliases&rdquo; column of <emu-xref href="#table-unicode-general-category-values"></emu-xref>, nor a binary property or binary property alias listed in the &ldquo;Property name and aliases&rdquo; column of <emu-xref href="#table-binary-unicode-properties"></emu-xref>.
           </li>
         </ul>
+        <emu-note>
+          <p>The semantics are extended in <emu-xref href="#sec-patterns-static-semantics-early-errors-annexb"></emu-xref>.</p>
```
Comment (michaelficarra) [inline spec.html:30841]: I'd replaces "The semantics are extended" with "This section is amended" everywhere:

```suggestion
          <p>This section is amended in <emu-xref href="#sec-patterns-static-semantics-early-errors-annexb"></emu-xref>.</p>
```

I'd also probably move each occurrence to the top of its respective section to make it more likely to be seen.

File: spec.html [hunk 13]
```diff
@@ -30748,11 +30748,14 @@ <h2>Syntax</h2>
       </emu-grammar>
 
       <emu-note>
-	      <p>A number of productions in this section are given alternative definitions in section <emu-xref href="#sec-regular-expressions-patterns"></emu-xref>.</p>
+        <p>A number of productions in this section are given alternative definitions in section <emu-xref href="#sec-regular-expressions-patterns"></emu-xref>.</p>
       </emu-note>
 
       <emu-clause id="sec-patterns-static-semantics-early-errors">
         <h1>Static Semantics: Early Errors</h1>
+          <emu-note>
```
Comment (michaelficarra) [inline spec.html:30756]: These notes should be indented at the same level as their siblings.

File: spec.html [hunk 14]
```diff
@@ -5171,7 +5177,8 @@ <h1>Abstract Equality Comparison</h1>
         1. If Type(_x_) is the same as Type(_y_), then
           1. Return the result of performing Strict Equality Comparison _x_ === _y_.
         1. If _x_ is *null* and _y_ is *undefined*, return *true*.
-        1. [id="step-abstract-equality-comparison-web-compat-insertion-point"] If _x_ is *undefined* and _y_ is *null*, return *true*.
+        1. If _x_ is *undefined* and _y_ is *null*, return *true*.
+        1. [id="step-abstract-equality-comparison-web-compat-insertion-point"] NOTE: Annex <emu-xref href="#sec-IsHTMLDDA-internal-slot-aec"></emu-xref> adds additional steps at this point.
```
Comment (michaelficarra) [inline spec.html:5181]: I'd prefer

```suggestion
        1. [id="step-abstract-equality-comparison-web-compat-insertion-point"] NOTE: This step is replaced in section <emu-xref href="#sec-IsHTMLDDA-internal-slot-aec"></emu-xref>.
```

to match the others. You'll have to reword the annex B section to match as well. @tc39/ecma262-editors what do you think?

Comment (ljharb) [reply spec.html:5181]: I'm fine with this suggestion.

Comment (daemon1024) [reply spec.html:5181]: @michaelficarra I believe the annex b reference already matches with it

**Review by michaelficarra — APPROVED** [review]
LGTM. Thanks so much!

**Review by ljharb — APPROVED** [review]

File: spec.html [hunk 15]
```diff
@@ -14308,6 +14316,9 @@ <h1>Runtime Semantics: Evaluation</h1>
             </tr>
             </tbody>
           </table>
+          <emu-note>
+            <p>Additional entry related to [[IsHTMLDDA]] Internal Slot can be found in <emu-xref href="#sec-IsHTMLDDA-internal-slot-typeof"></emu-xref>.</p>
```
Comment (syg) [inline spec.html:14320]:
```suggestion
            <p>An additional entry related to [[IsHTMLDDA]] Internal Slot can be found in <emu-xref href="#sec-IsHTMLDDA-internal-slot-typeof"></emu-xref>.</p>
```

**Review by syg — APPROVED** [review]

**Pull request merged**